src/osd/ReplicatedPG.cc

```diff
@@ -1496,6 +1496,22 @@ void ReplicatedPG::do_op(OpRequestRef& op)
     return;
   }
 
+  // discard due to cluster full transition?  (we discard any op that
+  // originates before the cluster or pool is marked full; the client
+  // will resend after the full flag is removed or if they expect the
+  // op to succeed despite being full).
+  // FIXME: we exclude mds writes for now.
+  if (!m->get_source().is_mds() &&
+      info.history.last_epoch_marked_full > m->get_map_epoch()) {
+    dout(10) << __func__ << " discarding op sent before full " << m << " "
+	     << *m << dendl;
+    return;
+  }
+  if (osd->check_failsafe_full()) {
+    osd->reply_op_error(op, -ENOSPC);
```

Comment:
Yeah, that seems like a good scheme. I think the mon would have to trust the osd's judgement on fullness, or there could be a race between increasing the setting on the mons and the osds going into the full state at the previous (lower) ratio before seeing the new settings. Putting the mon's thresholds in the osdmap rather than just the pgmap seems like a good idea though.

Could you clarify why you want to reject deletes when failsafe full is reached? Why not let the delete fail or succeed at a lower level?